akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala

```diff
@@ -703,6 +703,25 @@ final case class Referer(uri: Uri) extends jm.headers.Referer with RequestHeader
   def getUri: akka.http.javadsl.model.Uri = uri.asJava
 }
 
+//https://tools.ietf.org/html/rfc7231#section-7.1.3
+object `Retry-After` extends ModeledCompanion[`Retry-After`] {
+  def apply(delaySeconds: Long): `Retry-After` = apply(Left(delaySeconds))
+  def apply(timestamp: DateTime): `Retry-After` = apply(Right(timestamp))
+}
+final case class `Retry-After`(delaySecondsOrDateTime: Either[Long, DateTime]) extends jm.headers.RetryAfter with ResponseHeader {
```

Comment:
We would usually prefer modelling the data type explicitly instead of using the more general `Either`. The benefit is that user code will read nicer. How about adding an ADT like this:

```scala
sealed abstract class RetryAfter
final case class RetryAfterDuration(duration: Duration) extends RetryAfter
final case class RetryAfterTimestamp(timestamp: DateTime) extends RetryAfter
```

WDYT?

Reply:
In lack of a better option, the java version would still be fine.

Reply:
Moving to an ADT could work fine

Just to be clear, you expect the java supertype to still look the same or to do an equivalent operation there (splitting the java hierarchy too?).
I don't see any benefit there to be honest, but I may be missing some way to make it more usable in java too?

Reply:
@jrudolph I'm reading again your comment 'cause I'm not sure anymore that I correctly understood: is the ADT a substitute of the java class definition or just an implementation detail of the scala version?

I would keep the java version as is, because I'd avoid forcing users to do `instanceof` checks against the actual class, and would prefer the safer `Optional`-based api, in lack of a better alternative.
This is why I'm about to change only the scala implementation to encode it with the ADT in place of the internal `Either` member.

Reply:
Yeah, sorry if I wasn't clear. The suggestion was only for the Scala side. I don't have a better suggestion for the Java side so we can keep it this way.